### PR TITLE
feat: type ahead variable drop downs

### PR DESCRIPTION
**Pull request by jrenee42**
Closes #20195

behind the 'typeAheadVariableDropdown' feature flag

adds 'type-ahead' filtering for variable dropdowns in dashboards.

it used to look like this:
Screen Shot 2021-02-09 at 1 52 28 PM

it now looks like:
Screen Shot 2021-02-09 at 1 53 03 PM

behavior:
 * when you start typing, if the menu isn't already open it will open
 * when it first comes up, everything is shown (even if there is a pre-selected value); once you start typing tho it filters it down.  If you want to see everything in the list, then need to delete everything in the input field
 * when you leave the component, whatever value was selected will show up in the input; it won't let random strings show up
 * when you start typing, you can use the up and down arrows to make a selection.  if you make a selection and then click away, then the menu closes and you get that selection
 * if using the up/down arrows (or not), clicking 'return' sets the selection and closes the menu
    *if they were just typing, after pressing the return key the input field reverts to the selection that was last made (no random values are allowed)
 * it should be intuitive

- [x] [Well-formatted commit messages](https://www.conventionalcommits.org/en/v1.0.0-beta.3/)
- [x] Rebased/mergeable
- [x] Tests pass

### CODE REVIEW

**Comment by hercules-influx** [comment]
Metric | Score | Level
------------ | ------------- | ------------- 
performance | 45 | :red_circle:
accessibility | 65 | :yellow_circle:
best-practices | 87 | :green_circle:
seo | 80 | :yellow_circle:
pwa | null | :question:

File: src/variables/components/TypeAheadVariableDropdown.tsx [hunk 1]
```diff
@@ -0,0 +1,308 @@
...
+  onClickAwayHere = () => {
+    this.setState(this.getRealValue())
+  }
+
+  getRealValue = () => {
+    const {actualVal, typedValue} = this.state
+    const {selectedValue} = this.props
+
+    if (actualVal || selectedValue) {
+      const realValue = actualVal ? actualVal : selectedValue
```
Comment (genehynson) [inline src/variables/components/TypeAheadVariableDropdown.tsx:163]: you could shorten this to `const realValue = actualVal ?? selectedValue` if you wanted

File: src/variables/components/TypeAheadVariableDropdown.tsx [hunk 2]
```diff
@@ -0,0 +1,308 @@
...
+    const dropdownStatus =
+      values.length === 0 ? ComponentStatus.Disabled : ComponentStatus.Default
+
+    const placeHolderText = this.getPlaceHolderText('Select a Value')
+
+    const widthStyle = this.getWidth(placeHolderText)
+
+    return (
+      <Dropdown
+        style={{width: '140px'}}
```
Comment (genehynson) [inline src/variables/components/TypeAheadVariableDropdown.tsx:184]: could we move this into `variable-dropdown--dropdown` class?

File: src/variables/components/TypeAheadVariableDropdown.tsx [hunk 3]
```diff
@@ -0,0 +1,308 @@
...
+        menu={onCollapse => (
+          <Dropdown.Menu
+            style={widthStyle}
+            onCollapse={onCollapse}
+            theme={DropdownMenuTheme.Amethyst}
+          >
+            {shownValues.map(val => {
+              return (
+                <Dropdown.Item
+                  key={val}
```
Comment (genehynson) [inline src/variables/components/TypeAheadVariableDropdown.tsx:214]: do we know that all the values will be unique? If not, we could change the key to be the second `index` parameter of map instead

**Review by genehynson — APPROVED** [review]
lgtm! Left a couple small suggestions

File: src/variables/components/TypeAheadVariableDropdown.tsx (hunk 2 above)
Comment (jrenee42) [reply src/variables/components/TypeAheadVariableDropdown.tsx:184]: i'm worried about testing implications; and have already spend all week on this pr.....
this would be something for refactor-friday

File: src/variables/components/TypeAheadVariableDropdown.tsx (hunk 3 above)
Comment (jrenee42) [reply src/variables/components/TypeAheadVariableDropdown.tsx:214]: well, they are all strings and this is how it is now.  it's a good question; but don't want to change too much with this component change.

File: src/variables/components/TypeAheadVariableDropdown.tsx (hunk 1 above)
Comment (jrenee42) [reply src/variables/components/TypeAheadVariableDropdown.tsx]: done.

**Comment by jrenee42** [comment]
circle-ci tests all passed, despite what the issue says above.

File: src/variables/components/TypeAheadVariableDropdown.tsx (hunk 2 above)
Comment (genehynson) [reply src/variables/components/TypeAheadVariableDropdown.tsx:184]: I mean I highly doubt changing a css class would cause the tests to fail but ok...

**Comment by samhld** [comment]
Yaaaayyy -- awesome, @jrenee42 !
